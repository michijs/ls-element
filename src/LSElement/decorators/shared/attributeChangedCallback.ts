import { LSCustomElement } from '../../types';
import { getAttributeValue } from '../../utils/getAttribute';

export function attributeChangedCallback(self: LSCustomElement, name: string, oldValue, newValue) {
  if (newValue != oldValue) {
    self.componentWillReceiveAttribute?.(name, newValue, oldValue);
    self[name] = getAttributeValue(newValue);
  }
}