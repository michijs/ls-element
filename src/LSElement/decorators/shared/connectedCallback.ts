import type { LSCustomElement } from '../../types';
import { addReflectedAttributes } from '../../properties/addReflectedAttributes';
import { executeFirstRender } from '../../DOM/executeFirstRender';

export function connectedCallback(self: LSCustomElement) {//Children and attributes should be managed after constructor
  addReflectedAttributes(self);
  if (!self.ls.alreadyRendered) {
    self.componentWillMount?.();
    executeFirstRender(self);
    self.ls.alreadyRendered = true;
    self.componentDidMount?.();
  }
}