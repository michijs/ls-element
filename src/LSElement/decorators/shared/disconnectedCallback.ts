import { removeStoredAttributes } from '../../properties/removeStoredAttributes';
import type { LSCustomElement } from '../../types';

export function disconnectedCallback(self: LSCustomElement) {
  if (self.parentNode === null) {//TODO: search a better way to validate if element does not exists anymore
    removeStoredAttributes(self);
    self.componentDidUnmount?.();
  }
}