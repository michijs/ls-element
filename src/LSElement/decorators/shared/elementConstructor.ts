import type { LSCustomElement, ShadowOption, OtherShadowOptions } from '../../types';
import { addStores } from '../../properties/addStores';
import { initStore } from '../../properties/initStore';
import { addAttributes } from '../../properties/addAttributes';
import { addStoredAttributes } from '../../properties/addStoredAttributes';
import { addObservers } from '../../properties/addObservers';

export function elementConstructor(self: LSCustomElement, shadow: ShadowOption, otherShadowOptions: OtherShadowOptions) {
  if (shadow) {
    const shadowMode = shadow;
    const attachedShadow = self.attachShadow({ mode: shadowMode, ...otherShadowOptions });
    if (shadowMode === 'closed') {
      self._shadowRoot = attachedShadow;
    }
  }
  self.ls = self.ls || {};
  initStore(self);
  addStores(self);
  addAttributes(self);
  addStoredAttributes(self);
  addObservers(self);
}