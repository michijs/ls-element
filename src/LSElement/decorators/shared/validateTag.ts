export const validateTag = (tag: string) => {
  if (tag.indexOf('-') <= 0) {
    throw new Error('You need at least 1 dash in the custom element name!');
  }
};
// TODO: ver si vale la pena