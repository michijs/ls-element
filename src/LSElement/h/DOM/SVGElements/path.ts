import { AllAttributes } from '../DOMAttributes/AllAttributes';
import { SVGAriaAttributes, SVGConditionalProcessingAttributes, SVGGenericAttributes, SVGPresentationAttributes } from '../DOMAttributes/SVG';
import { SVGEvents } from '../DOMEvents/SVGEvents';

export interface path extends Partial<
    Pick<AllAttributes, 'd' | 'pathLength'>
    & SVGGenericAttributes
    & SVGConditionalProcessingAttributes
    & SVGEvents
    & SVGPresentationAttributes
    & SVGAriaAttributes
>{}