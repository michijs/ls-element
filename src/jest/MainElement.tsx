import { Attribute, AutonomousCustomElement, CustomElementWrapper, h, LSCustomElement } from '../index';

function random(max) {
  return Math.round(Math.random() * 1000) % max;
}

const A = ['pretty', 'large', 'big', 'small', 'tall', 'short', 'long', 'handsome', 'plain', 'quaint', 'clean',
  'elegant', 'easy', 'angry', 'crazy', 'helpful', 'mushy', 'odd', 'unsightly', 'adorable', 'important', 'inexpensive',
  'cheap', 'expensive', 'fancy'];
const C = ['red', 'yellow', 'blue', 'green', 'pink', 'brown', 'purple', 'brown', 'white', 'black', 'orange'];
const N = ['table', 'chair', 'house', 'bbq', 'desk', 'car', 'pony', 'cookie', 'sandwich', 'burger', 'pizza', 'mouse',
  'keyboard'];

@AutonomousCustomElement({ shadow: false })
export class MainElement extends HTMLElement implements LSCustomElement {
  @Attribute() rows = [];
  @Attribute() selected = 0;
  nextId = 1;

  buildData(count) {
    const data = new Array(count);
    for (let i = 0; i < count; i++) {
      data[i] = {
        id: this.nextId++,
        label: `${A[random(A.length)]} ${C[random(C.length)]} ${N[random(N.length)]}`,
      };
    }
    return data;
  }

  render() {
    return (
      <div id="container" _className="container" _dynamicAttributes={[]}>
        <div id="jumbotron" _className="jumbotron" _dynamicAttributes={[]} _staticChildren>
          <div id="1" _className="row">
            <div id="2" _className="col-md-6">
              <h1 id="3" _textContent="LS-Element keyed" />
            </div>
            <div id="4" _className="col-md-6">
              <div id="5" _className="row">
                <div id="6" _className="col-sm-6 smallpad">
                  <button _type="button" _className="btn btn-primary btn-block" id="run" onclick={this.run} _textContent="Create 1,000 rows" />
                </div>
                <div id="7" _className="col-sm-6 smallpad">
                  <button _type="button" _className="btn btn-primary btn-block" id="runlots" onclick={this.runLots} _textContent="Create 10,000 rows" />
                </div>
                <div id="8" _className="col-sm-6 smallpad">
                  <button _type="button" _className="btn btn-primary btn-block" id="add" onclick={this.add} _textContent="Append 1,000 rows" />
                </div>
                <div id="9" _className="col-sm-6 smallpad">
                  <button _type="button" _className="btn btn-primary btn-block" id="update" onclick={() => this.update()} _textContent="Update every 10th row" />
                </div>
                <div id="10" _className="col-sm-6 smallpad">
                  <button _type="button" _className="btn btn-primary btn-block" id="clear" onclick={this.clear} _textContent="Clear" />
                </div>
                <div id="11" _className="col-sm-6 smallpad">
                  <button _type="button" _className="btn btn-primary btn-block" id="swaprows" onclick={this.swapRows} _textContent="Swap Rows" />
                </div >
              </div >
            </div >
          </div >
        </div >
        <table id="table" _className="table table-hover table-striped test-data" _dynamicAttributes={[]}>
          <tbody id="tbody" _dynamicAttributes={[]}>{this.rows.map(({ id, label }) => (
            <tr id={`row-${id}`} class={id === this.selected ? 'danger' : undefined} _dynamicAttributes={['class']}>
              <td id={`row-${id}-1`} _className="col-md-1" _dynamicAttributes={[]} _staticChildren _textContent={id} />
              <td id={`row-${id}-2`} _className="col-md-4" _dynamicAttributes={[]}>
                <a id={`row-${id}-3`} onclick={() => this.select(id)} _staticChildren _dynamicAttributes={['_textContent']} _textContent={label} />
              </td>
              <td id={`row-${id}-4`} _className="col-md-1" _dynamicAttributes={[]} _staticChildren>
                <a id={`row-${id}-5`} onclick={() => this.delete(id)}>
                  <span id={`row-${id}-6`} _className="glyphicon glyphicon-remove" aria-hidden="true" />
                </a>
              </td>
              <td id={`row-${id}-7`} _className="col-md-6" _dynamicAttributes={[]} _staticChildren />
            </tr>))}
          </tbody>
        </table >
        <span id="icon" _className="preloadicon glyphicon glyphicon-remove" aria-hidden="true" _dynamicAttributes={[]} _staticChildren />
      </div >
    );
  }
  run() {
    this.rows = this.buildData(1000);
    this.selected = 0;
  }
  runLots() {
    this.rows = this.buildData(10000);
    this.selected = 0;
  }
  add() {
    this.rows = this.rows.concat(this.buildData(1000));
  }
  delete(id) {
    this.rows = this.rows.filter(x => x.id !== id);
  }
  select(id) {
    this.selected = id;
  }
  update(mod = 10) {
    this.rows = this.rows.map((row, i) => {
      return i % mod === 0 ? { ...row, label: `${row.label} !!!` } : row;
    });
  }
  clear() {
    this.rows = [];
    this.selected = 0;
  }
  swapRows() {
    if (this.rows.length > 998) {
      const d1 = this.rows[1];

      this.rows = this.rows.map((data, i) => {
        if (i === 1) {
          return this.rows[998];
        }
        else if (i === 998) {
          return d1;
        }
        return data;
      });
    }
  }
}

export default CustomElementWrapper()(MainElement);