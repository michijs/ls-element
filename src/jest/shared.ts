import { MainElement } from './MainElement';

export type Result =
    'create1000Rows'
    | 'replaceAllRows'
    | 'partialUpdate'
    | 'selectRow'
    | 'swapRows'
    | 'removeRow'
    | 'createManyRows'
    | 'appendRowsToLargeTable'
    | 'clearRows'

const create1000Rows = () => {
  document.getElementById('run').click();
};
const add1000Rows = () => {
  document.getElementById('add').click();
};
const create10000Rows = () => {
  document.getElementById('runlots').click();
};
const updateEvery10Rows = () => {
  document.getElementById('update').click();
};
const swapRows = () => {
  document.getElementById('swaprows').click();
};
const getTableBody = () => {
  return Array.from(document.getElementById('tbody').children);
};
const clear = () => {
  document.getElementById('clear').click();
};
const getRowId = (element: Element) => {
  return Number(element.childNodes.item(0).childNodes.item(0).textContent);
};

export function makePerformanceTests(dataController: () => MainElement, selectByIndex: boolean) {
  const results = new Map<Result, number>();
  const saveResult = (key: Result, functionToMeasure: () => void) => {
    const t0 = performance.now();
    functionToMeasure();
    const t1 = performance.now();
    results.set(key, Number(Number(t1 - t0).toFixed(2)));
  };
  it('creates 1000 rows when clicking run', () => {
    saveResult('create1000Rows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
  });
  it('replaces 1000 rows when clicking run', () => {
    create1000Rows();
    saveResult('replaceAllRows', create1000Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach(element => {
      expect(getRowId(element)).toBeGreaterThan(1000);
    });
  });
  it('update every 10th row 1000 rows on a table with 1000 rows when clicking update', () => {
    create1000Rows();
    saveResult('partialUpdate', updateEvery10Rows);
    expect(getTableBody().length).toEqual(1000);
    getTableBody().forEach((element, index) => {
      if (index % 10 === 0) {
        expect(element.innerHTML.includes('!!!')).toBeTruthy();
      } else {
        expect(element.innerHTML.includes('!!!')).toBeFalsy();
      }
    });
  });
  it('select a row (1000 rows)', () => {
    create1000Rows();
    saveResult('selectRow', () => dataController().select(999));
    expect(getTableBody()[selectByIndex ? 999 : 998].className).toEqual('danger');
  });
  it('swap a row (1000 rows)', () => {
    create1000Rows();
    saveResult('swapRows', swapRows);
    expect(getRowId(getTableBody()[1])).toEqual(999);
    expect(getRowId(getTableBody()[998])).toEqual(2);
  });
  it('remove a row (1000 rows)', () => {
    create1000Rows();
    saveResult('removeRow', () => dataController().delete(999));
    expect(getTableBody().length).toEqual(999);
  });
  it('creates 10000 rows when clicking runlots', () => {
    saveResult('createManyRows', create10000Rows);
    expect(getTableBody().length).toEqual(10000);
  });
  it('append 1000 rows on a large table', () => {
    create10000Rows();
    saveResult('appendRowsToLargeTable', add1000Rows);
    expect(getTableBody().length).toEqual(11000);
  });
  it('clear rows', () => {
    create1000Rows();
    saveResult('clearRows', clear);
    expect(getTableBody().length).toEqual(0);
  });
  return results;
}