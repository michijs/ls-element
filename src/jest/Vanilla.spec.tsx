import { MainElement } from './MainElement';
import { makePerformanceTests } from './shared';
import { Main as vanillaDataController } from './vanillajs/vanillajs.js';
import * as fs from 'fs';
import * as path from 'path';
const vanillaTemplate = fs.readFileSync(path.resolve(__dirname, './vanillajs/index.html'), 'utf8');


describe('Performance tests - vanilla-js', () => {
  let component: MainElement;
  beforeEach(() => {
    document.body.innerHTML = '';
    const html = document.createElement('div');
    html.innerHTML += vanillaTemplate;
    document.body.appendChild(html);
    component = new vanillaDataController() as unknown as MainElement;
  });
  const getComponent = () => {
    return component;
  };

  const results = makePerformanceTests(getComponent, true);
  afterAll(() => {
    expect(results).toMatchSnapshot('vanilla JS');
  });
});