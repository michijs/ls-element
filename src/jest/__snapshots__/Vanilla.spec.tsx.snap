// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance tests - vanilla-js clear rows: Vanilla JS 1`] = `
Map {
  "create1000Rows" => 157.92,
  "replaceAllRows" => 167.2,
  "partialUpdate" => 69.45,
  "selectRow" => 38.8,
  "swapRows" => 44.47,
  "removeRow" => 21.75,
  "createManyRows" => 1200.53,
  "appendRowsToLargeTable" => 215.77,
  "clearRows" => 36.44,
}
`;
